### DISCUSSION TITLE
Production build broken due to missing css

### DISCUSSION THREAD

**Discussion by josemigallas** [webpack/webpack#18808] [Q&A]
I've spent days with this issue and haven't figured it out yet so I thought maybe the good peoples of webpack could lend me a hand.

The project is a huge one made with rails and we use webpack to compile Typescript/React and to bundle libraries from node_modules. The compiled files are listed in a `manifest.json` and a helper method from rails then append these in the HTML templates. CSS is extracted to separate files with `mini-css-extract-plugin`.

For the CSS we rely on a 3rd party library, Patternfly, that exports React components shipped with the necessary stylesheet.

In development (running webpack-dev-server) everything works and looks as it should, chunks and everything. But then in production, some of these stylesheets that should come with the React components are shaken off from the chunks:
<img width="1162" alt="Screenshot 2024-09-27 at 14 43 35" src="https://github.com/user-attachments/assets/10f18ef0-7ac4-42aa-9615-89523c4ab773">

There are no errors or warnings, the stylesheets are simply not included in the "bundle".

Comparing development vs production bundles, here are some relevant lines involving the aforementioned stylesheet:

### Development

```js
// Chunk: vendors-node_modules_patternfly_quickstarts_dist_index_es_js-node_modules_patternfly_react-co-ID.js
/* ... */

/* harmony import */ var _patternfly_react_core__WEBPACK_IMPORTED_MODULE_48__ = __webpack_require__(/*! @patternfly/react-core */ "./node_modules/@patternfly/react-core/dist/esm/components/EmptyState/EmptyState.js");

/* ... */

/***/ "./node_modules/@patternfly/react-core/dist/esm/components/EmptyState/EmptyState.js":
/*!******************************************************************************************!*\
  !*** ./node_modules/@patternfly/react-core/dist/esm/components/EmptyState/EmptyState.js ***!
  \******************************************************************************************/
/***/ ((__unused_webpack_module, __webpack_exports__, __webpack_require__) => {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   EmptyState: () => (/* binding */ EmptyState),
/* harmony export */   EmptyStateVariant: () => (/* binding */ EmptyStateVariant)
/* harmony export */ });
/* harmony import */ var tslib__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(/*! tslib */ "./node_modules/tslib/tslib.es6.js");
/* harmony import */ var react__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! react */ "./node_modules/react/index.js");
/* harmony import */ var _patternfly_react_styles__WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(/*! @patternfly/react-styles */ "./node_modules/@patternfly/react-styles/dist/esm/index.js");
/* harmony import */ var _patternfly_react_styles_css_components_EmptyState_empty_state__WEBPACK_IMPORTED_MODULE_3__ = __webpack_require__(/*! @patternfly/react-styles/css/components/EmptyState/empty-state */ "./node_modules/@patternfly/react-styles/css/components/EmptyState/empty-state.js");




var EmptyStateVariant;
(function (EmptyStateVariant) {
    EmptyStateVariant["xs"] = "xs";
    EmptyStateVariant["small"] = "small";
    EmptyStateVariant["large"] = "large";
    EmptyStateVariant["xl"] = "xl";
    EmptyStateVariant["full"] = "full";
})(EmptyStateVariant || (EmptyStateVariant = {}));
const EmptyState = (_a) => {
    var { children, className = '', variant = EmptyStateVariant.full, isFullHeight } = _a, props = (0,tslib__WEBPACK_IMPORTED_MODULE_1__.__rest)(_a, ["children", "className", "variant", "isFullHeight"]);
    return (react__WEBPACK_IMPORTED_MODULE_0__.createElement("div", Object.assign({ className: (0,_patternfly_react_styles__WEBPACK_IMPORTED_MODULE_2__.css)(_patternfly_react_styles_css_components_EmptyState_empty_state__WEBPACK_IMPORTED_MODULE_3__["default"].emptyState, variant === 'xs' && _patternfly_react_styles_css_components_EmptyState_empty_state__WEBPACK_IMPORTED_MODULE_3__["default"].modifiers.xs, variant === 'small' && _patternfly_react_styles_css_components_EmptyState_empty_state__WEBPACK_IMPORTED_MODULE_3__["default"].modifiers.sm, variant === 'large' && _patternfly_react_styles_css_components_EmptyState_empty_state__WEBPACK_IMPORTED_MODULE_3__["default"].modifiers.lg, variant === 'xl' && _patternfly_react_styles_css_components_EmptyState_empty_state__WEBPACK_IMPORTED_MODULE_3__["default"].modifiers.xl, isFullHeight && _patternfly_react_styles_css_components_EmptyState_empty_state__WEBPACK_IMPORTED_MODULE_3__["default"].modifiers.fullHeight, className) }, props),
        react__WEBPACK_IMPORTED_MODULE_0__.createElement("div", { className: (0,_patternfly_react_styles__WEBPACK_IMPORTED_MODULE_2__.css)(_patternfly_react_styles_css_components_EmptyState_empty_state__WEBPACK_IMPORTED_MODULE_3__["default"].emptyStateContent) }, children)));
};
EmptyState.displayName = 'EmptyState';
//# sourceMappingURL=EmptyState.js.map

/***/ }),

/* ... */
```

```css
// Chunk: vendors-node_modules_patternfly_quickstarts_dist_index_es_js-node_modules_patternfly_react-co-ID.css
/* ... */

/*!***************************************************************************************************************************************************************!*\
  !*** css ./node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[1].use[1]!./node_modules/@patternfly/react-styles/css/components/EmptyState/empty-state.css ***!
  \***************************************************************************************************************************************************************/
.pf-c-empty-state {

/* ... */
```

### Production

```js
// Chunk: vendors-node_modules_patternfly_quickstarts_dist_index_es_js-node_modules_patternfly_react-co-ID.js

/* ... */

// UNUSED EXPORTS: /* ... */ QuickStartCatalogEmptyState, /* ... */

/* ... */

const QuickStartCatalogEmptyState = ({ clearFilters }) => {
    const { getResource } = React.useContext(QuickStartContext);
    return (React.createElement(EmptyState, null, // <---- Symbol used here?
        React.createElement(EmptyStateIcon, { icon: SearchIcon }),
        React.createElement(Title, { size: "lg", headingLevel: "h4" }, getResource('No results found')),
        React.createElement(EmptyStateBody, null, getResource('No results match the filter criteria. Remove filters or clear all filters to show results.')),
        React.createElement(EmptyStatePrimary, null,
            React.createElement(Button, { variant: "link", onClick: clearFilters, "data-test": "clear-filter button" }, getResource('Clear all filters')))));
};

/* ... */
```

```css
// Chunk: vendors-node_modules_patternfly_quickstarts_dist_index_es_js-node_modules_patternfly_react-co-ID.css

// No signs of "empty-state" ! 🤷‍♂️
```

Any ideas? How could I further debug this? I am clueless.

Webpack config: https://github.com/3scale/porta/pull/3848/files#diff-2d9f41ba15e62121a6231d24822574b7a743869ccc9f53e7a79259222e1f915f (split into base, development, production, rules and plugins)

**Comment by alexander-akait** [comment]
Can you provide steps to reproduce? All problems with an unfilled issue template automatically go to discussions, but this does not mean that I will not help solve the problem.

Possible problem is that some packages are marked `sideEffects: false`, so webpack just removes them.

Another problem may be - https://github.com/3scale/porta/pull/3848/files#diff-b831107fab6c7ffa5c1c2916fab214f18c6320cef90f1888d4c0ff142026ac3bR22, we have an official solution - https://github.com/webpack-contrib/terser-webpack-plugin?tab=readme-ov-file#esbuild, I also recommend trying to use the built-in minimizer (i.e. avoid `minify` and webpack try to minify it using terser).

And also try to remove - https://github.com/3scale/porta/pull/3848/files#diff-ff09658f63a0d1d979b833bebe1536ad9c10341e2cd624426986ff8d8a580c41R16 - Some chunks are still needed to understand that they are loaded

Feel free to feedback

**Reply by josemigallas** [reply]
Hi @alexander-akait thanks for the quick response.

> Can you provide steps to reproduce? All problems with an unfilled issue template automatically go to discussions, but this does not mean that I will not help solve the problem.

I'm not sure there is a way to reproduce this other than pulling the repo, setting up the project and running it yourself 

> Possible problem is that some packages are marked `sideEffects: false`, so webpack just removes them.

I ruled that out because stylesheets are ultimately imported from `@patternfly/react-core` and it has sideEffects:
https://github.com/patternfly/patternfly-react/blob/db0cdde6e3e6ae61348129b73bb43c97a98b4617/packages/react-core/package.json#L19-L22

> Another problem may be - https://github.com/3scale/porta/pull/3848/files#diff-b831107fab6c7ffa5c1c2916fab214f18c6320cef90f1888d4c0ff142026ac3bR22, we have an official solution - https://github.com/webpack-contrib/terser-webpack-plugin?tab=readme-ov-file#esbuild, I also recommend trying to use the built-in minimizer (i.e. avoid `minify` and webpack try to minify it using terser).

I tried removing this, also adding it to development and it doesn't make any difference :(

> And also try to remove - https://github.com/3scale/porta/pull/3848/files#diff-ff09658f63a0d1d979b833bebe1536ad9c10341e2cd624426986ff8d8a580c41R16 - Some chunks are still needed to understand that they are loaded

In fact I added this after, as a possible solution.

Ideally what I would like is that in development mode the styles are also missing, not the other way around. Is there any way I can force production-like tree shaking?

**Reply by alexander-akait** [reply]
> I'm not sure there is a way to reproduce this other than pulling the repo, setting up the project and running it yourself

Can you provide steps to reproduce? I will look

**Reply by josemigallas** [reply]
> Can you provide steps to reproduce? I will look

If you are able to run the project locally, by following [INSTALL.md](https://github.com/3scale/porta/blob/master/INSTALL.md#install), the it is reproducible by comparing `yarn dev` vs. `yarn build`. The first runs webpack-dev-server in mode development, the second runs webpack in mode production.

I am trying to create a minimal repo that reproduces this error (just webpack and React) but that's raising its own issues 😅.

The only workaround I can think of is to run webpack-dev-server in production mode... or run webpack with `--watch`. We will lose some development features for sure, but it's better than a "false positive" development.

**Reply by josemigallas** [reply]
@alexander-akait I found the configs that are breaking my production build:
```js
webpackConfig.optimization.usedExports = true
webpackConfig.optimization.innerGraph = true
```

Both in combination remove the css chunks. I will set this as default so that my development bundle is accurate compared to production.

**Reply by alexander-akait** [reply]
Using this you are destroying a lot of optimizations, I strongly recommend to avoid this and find the real reason

**Reply by alexander-akait** [reply]
> If you are able to run the project locally, by following [INSTALL.md](https://github.com/3scale/porta/blob/master/INSTALL.md#install), the it is reproducible by comparing yarn dev vs. yarn build. The first runs webpack-dev-server in mode development, the second runs webpack in mode production.

I don't see any `yarn dev` and `yarn build`

Please provide the working steps and how can I debug and see the problem

Even more - I see you use webpack@4 - https://github.com/3scale/porta/blob/master/package.json#L42

I'm almost sure that the problem is in `sideEffect`

Try this - https://webpack.js.org/configuration/module/#rulesideeffects

For all JS and CSS loaders to debug it

**Reply by josemigallas** [reply]
> Using this you are destroying a lot of optimizations, I strongly recommend to avoid this and find the real reason

Isn't it the other way around? I am adding these 2 optimization options to the development build. If anything, I am adding unnecessary optimization. Am I wrong?
https://github.com/3scale/porta/blob/bdde485fad9e3234bbcc882f1090dbf5b83d9a88/config/webpack/development.config.js#L42-L48

> I don't see any yarn dev and yarn build
> [...] 
> Even more - I see you use webpack@4 -

Apologies, I pasted the one from `master`, however I am working on branch [bye_bye_webpacker](https://github.com/3scale/porta/blob/bye_bye_webpacker/package.json) that uses webpack@5.

> I'm almost sure that the problem is in sideEffect

Maybe I missed something, but I added one by one all optimization defaults from https://github.com/webpack/webpack/blob/8fb481d9dbbdbb1be0c14028c9a1c3ce3b66b5bd/lib/config/defaults.js#L1415 to the development config and the ones that made it looks like production were the ones above: `usedExports` and `innerGraph`.


> Possible problem is that some packages are marked sideEffects: false, so webpack just removes them.

Coming back to the start of the discussion, I think the real problem is that development config includes stylesheets that should not be there, rather than production missing them. I think these stylesheets are bundled, coming from React components that **are not used** in that particular page and therefore it's unused exports, dead code. So it's the development config what's wrong and that is why I added those optimization flags. This way, when I am developing a new feature, I see the same as production and I don't have to verify both builds to find missing chunks.

Does that make sense to you @alexander-akait ?

**Reply by josemigallas** [reply]
> Please provide the working steps and how can I debug and see the problem

Here are details steps but I must reiterate the difficult part is to get the application running... 😞 

1. Run the application from branch [`bye_bye_webpacker`](https://github.com/3scale/porta/tree/bye_bye_webpacker) by following INSTALL.md

2.  Run `yarn dev` in a different terminal.

3. Open in browser http://provider-admin.3scale.localhost:3000/p/admin/account/authentication_providers/new, it should look like: 
![Screenshot 2024-10-04 at 12 10 52](https://github.com/user-attachments/assets/eb327dbe-1425-48a7-9e75-9ce2468a7676)

4. Remove the [following lines](https://github.com/3scale/porta/blob/bdde485fad9e3234bbcc882f1090dbf5b83d9a88/app/views/provider/admin/account/authentication_providers/new.html.slim#L3-L4) to not import the stylesheet explicitly. It should look like: 
![Screenshot 2024-10-04 at 12 16 59](https://github.com/user-attachments/assets/986f086e-c14d-49ef-9d67-3f282a8d0046)

5. Remove the [2 optimization configs from development](https://github.com/3scale/porta/blob/bdde485fad9e3234bbcc882f1090dbf5b83d9a88/config/webpack/development.config.js#L46-L47). Restart `yarn dev` and it again should look like: ![Screenshot 2024-10-04 at 12 10 52](https://github.com/user-attachments/assets/eb327dbe-1425-48a7-9e75-9ce2468a7676)

6. Stop `yarn dev` and run `yarn build` (production) and refresh the browser, now it shoud look like: ![Screenshot 2024-10-04 at 12 16 59](https://github.com/user-attachments/assets/986f086e-c14d-49ef-9d67-3f282a8d0046)

**Reply by alexander-akait** [reply]
@josemigallas Sorry for the delay, there were a lot of tasks, I see the branch was deleted, have the problems been resolved? Was everything resolved without any workarounds?

**Reply by josemigallas** [reply]
Hi @alexander-akait, the problem wasn't solved but worked around. Our team need to make prod and dev builds equal so we ended up enabling the 2 options there were causing the inconsistency:
https://github.com/3scale/porta/commit/1e0a1a23808745bd59bc5b4f16f03c29fa4ec9de#diff-7fc4b97d24be7226d6a402bf7c133b390797f0863f416b3ea72959e2d67fdf87R42-R48

Once we did this, the styles were also missing in Development and therefore we were able to add the necessary import statements so that we don't stumble upon a broken UI in production.

**Reply by alexander-akait** [reply]
@josemigallas Could you create a repository without extra code as it requires a lot of setup? I'm almost sure that the problem is in some packages related to incorrect side effects settings.

**Reply by josemigallas** [reply]
@alexander-akait Apologies for the late response, I am currently on leave and I barely have time to do spend on the laptop. I will revisit this as soon as I can and try to create a small repo to reproduce.

**Comment by fregante** [comment]
I had the same issue and it was resolved by removing `sideEffects: false` from the imported package.

It seems that the affected package has an array in that key, which potentially is interpreted as a truthy value by mini-css-extract-plugin:

https://github.com/patternfly/patternfly-react/blame/f8b21b594c90960305e3b58b723c1cff7a9decbb/packages/react-styles/package.json#L7-L10

I think this could be classified as a "bug" in both packages:

- the package cannot have `sideEffects` because it has CSS
- webpack should not trim CSS just because the imported package has `sideEffects`

**Reply by alexander-akait** [reply]
@fregante We respect `sideEffects`, if your CSS was removed it means your have wrong value of `sideEffects` in your `package.json`, note - any JS file that import CSS is not side effect free, it is DOM modification, like if your JS module have `document.head.append(style)`, your literally inject styles tags and it has side effects, yes, in most cases this will not create problems because you are using named export like `import { style } from "./style.css"` and has `sideEffect: ["*.css"]` we see it is a CSS file and export class and has side effects, but if you have `import "./style.css"` you can have problems
